contrib/seccomp-notify-plugin-example-rust/src/seccompwatcher.rs

```diff
@@ -0,0 +1,244 @@
+        Ok(s) => {
+            s
+        }
+        Err(_) => {
+            return Err(nc::EINVAL);
+        }
+    };
+    let filename = match f.into_string() {
+        Ok(s) => {
+            s
+        }
+        Err(_) => {
+            return Err(nc::EINVAL);
+        }
+    };
+    Ok(filename)
+}
+
+fn handle_mknodat_request(req: &mut nc::seccomp_notif_t) -> Result<u32, nc::Errno> {
+    return handle_mknod_internal(0, 1, 2, 3, req)
+}
+
+fn handle_mknod_request(req: &mut nc::seccomp_notif_t) -> Result<u32, nc::Errno> {
+    return handle_mknod_internal(-1, 0, 1, 2, req)
+
+}
+
+fn get_file_from_fd(fd: std::os::unix::io::RawFd) -> std::fs::File {
+    let r = unsafe {std::fs::File::from_raw_fd(fd)};
+    r
+}
+
+fn handle_mknod_internal(dirfd: isize, path_arg: usize, mode_arg: usize, dev_arg: usize, req: &mut nc::seccomp_notif_t) -> Result<u32, nc::Errno> {
+    let mut fname_buffer = vec![0; nc::types::PATH_MAX as usize];
+
+    let fd = get_file_from_fd(nc::open(&format!("/proc/{}/mem", req.pid), nc::O_RDONLY, 0)?);
+
+    nc::pread64(fd.as_raw_fd(), &mut fname_buffer, req.data.args[path_arg] as isize)?;
+
+    let filename = get_string_from_buffer(fname_buffer)?;
+
+    // Only char devices are allowed
+    if (req.data.args[mode_arg] as u32 & nc::S_IFMT) != nc::S_IFCHR {
+        return Err(nc::EPERM);
+    }
+    let mode = (req.data.args[mode_arg] & 01777) as u32;
+
+    let major = ((req.data.args[dev_arg] >> 8) & 0xFF) as u32;
+    let minor = (req.data.args[dev_arg] & 0xFF) as u32;
+    let present: &Device = match ALLOWED_DEVICES.iter().find(|&x| x.major == major && x.minor == minor) {
+        Some(x) => x,
+        None => {
+            return Err(nc::EPERM);
+        }
+    };
+
+    let mut hdr = nc::cap_user_header_t{
+        version: nc::LINUX_CAPABILITY_VERSION_3 as u32,
+        pid: req.pid as i32,
+    };
+    let mut udata = nc::cap_user_data_t{
+        permitted: 0,
+        effective: 0,
+        inheritable: 0};
+    nc::capget(&mut hdr, &mut udata)?;
+
+    // The process has no CAP_MKNOD
+    let has_cap_mknod = udata.effective & (1 << nc::types::CAP_MKNOD) != 0;
+    if !has_cap_mknod {
+        return Err(nc::EPERM);
+    }
+
+    let pid = nc::fork()?;
+    if pid == 0 {
+        let cwd_fd: std::fs::File = if dirfd < 0 {
+            get_file_from_fd(nc::open(&format!("/proc/{}/cwd", req.pid), nc::O_RDONLY, 0)?)
+        } else {
+            get_file_from_fd(nc::open(&format!("/proc/{}/fd/{}", req.pid, req.data.args[dirfd as usize]), nc::O_RDONLY, 0)?)
+        };
+        match do_mount_in_mountns (cwd_fd.as_raw_fd(), &filename, mode, &present.from.to_string(), req) {
+            Ok(_) => {
+                nc::exit(0);
+            }
+            Err(errno) => {
+                nc::exit(errno as u8);
+            }
+        }
+    }
+
+    let mut usage: nc::rusage_t  = unsafe {std::mem::zeroed()};
+    let mut status: i32 = 0;
+    nc::wait4(pid, &mut status, 0, &mut usage)?;
```

Comment:
And `wait4` is [in libc](https://docs.rs/libc/0.2.74/libc/fn.wait4.html) - seems simpler to use that than this `nc` wrapper.